### DISCUSSION TITLE
"long-term" C++14 support branch based on 3.6.0?

### DISCUSSION THREAD

**Discussion by insi-eb** [CppMicroServices/CppMicroServices#701] [Ideas]
we are using CppMicroServices in an environment where at least in the next year or so, we cannot move to C++17.

Is there any interest for including a C++14 support branch based on 3.6.0 in this repo, if we would step in and do the work to backport important bugfixes (where possible) and update 3rd-party dependencies to such a branch? Or should we just fork the repo and do what we need on our fork?

**Comment by jeffdiclemente** [comment]
@insi-eb , absolutely we are ok with a C++14 support branch. If you are willing to backport the fixes you need in a fork, we can create a branch in this repo for you to merge into from your fork. There are probably others who would benefit from a C++14 support branch so having it generally available on a dedicated branch in this repo would be beneficial. 

Let me know if this is acceptable to you.

**Reply by achristoforides** [reply]
I wanted to add a bit of information to what @jeffdiclemente suggested regarding maintaining a dedicated c++14 support branch.

All commits up to and including the `v3.6.0` tag on master are c++14 compliant. Additionally, all commits from that point (v3.6.0 equivalent commit on `development`) forward up to and including https://github.com/CppMicroServices/CppMicroServices/commit/d955a4296ee7f937da1a6ed26372de0c735266a8 are similarly c++14 compliant. The commit _after_ the one I linked previously (`d955a42`) is the actual commit on `development` to switch the project to use c++17.

At a minimum, most of the changes from August 13, 2020 all the way through March 25, 2022 on the `development` branch can be simply ported by replaying the commits of `development` starting at the commit after the `v3.6.0` tag and ending at the commit I linked above. Once this is complete, the number of changes to port would be dramatically smaller.

**Reply by insi-eb** [reply]
@jeffdiclemente yes, that setup sounds good for us
@achristoforides thanks, that info is helpful

**Reply by jeffdiclemente** [reply]
@insi-eb , I've created the [c++14-compliant](https://github.com/CppMicroServices/CppMicroServices/tree/c++14-compliant) branch which you can use for pull requests. This branch was created from https://github.com/CppMicroServices/CppMicroServices/commit/d955a4296ee7f937da1a6ed26372de0c735266a8.

**Reply by insi-eb** [reply]
how would the preferred workmode be for porting bugfixes/changes and merging things back to the project? I've looked at your Contribution guidelines, but I believe they are not really covering this case.

If there are no changes needed to get a change from master (or later develop) to be built in C++14 mode, can I just cherry pick the corresponding commit onto the c++14-compliant branch of our fork and get a bunch of those merged together from the fork to the project? or what would you prefer?
If there are small changes (e.g. additional includes/extract variable initialization from if blovk) needed, is it ok to cherry-pick and add the changes in a separate commit or should I do everything in a completely new commit? I could merge that still inside the fork from a branch named "701-backport-xxx-to-c++14" to keep to the existing naming scheme ...

I guess if I need larger changes for anything, it would make sense to get a review from the project before I even merge it to the c++14-compliant branch of our fork.

**Reply by achristoforides** [reply]
@insi-eb I think mimicking the strategy we use for contributions to the repo should apply in this case, albeit with some modification.

To answer your first question: if you identify a change on master or development that when ported to `c++14-compliant` does not require any modification, I would still recommend following the steps I describe below (with the cherry-picking addendum to step (2)) as I think that is the path of least resistance when porting back the changes to the `c++14-compliant` branch in this repo.

My suggestion would be the following:

If back-porting a change to the `c++14-compliant`:
1. Identify a change from `development/CppMicroServices` that you believe should be back-ported to the `c++14-compliant` branch.
2. Create a branch off of `c++14-compliant` in your fork (titled however you deem appropriate) and do your development work on that branch.
   - If the change does not require any modifications to work in the `c++14-compliant` branch, you should be able to cherry pick the commit from master/development into the branch you just created.
3. When you're ready for it to be reviewed/merged, create a PR in this repo where the branches selected are "`c++14-compliant` <- `<your_branch_from_your_fork>`".
4. Once it has been reviewed, we will merge it into the `c++14-compliant` branch in this repo and you should be able to reflect those changes in your fork's `c++14-compliant` branch by simply pulling from the upstream `c++14-compliant` branch (aka the one in this repo) into your fork's `c++14-compliant` branch.
5. Rinse and repeat.

In terms of upstreaming new code/fixes/features that you author into `development` in this repo, you should simply be able to make branches off of your up-to-date-with-upstream `c++14-compat` branch in your fork, author your change, and create a PR in this repo using that branch. If this workflow runs into any issues, please let us know and we can try to figure out a better solution.

Does this all make sense and would this work for you?

**Reply by insi-eb** [reply]
yes, it makes sense - I now just have to wait for my company's OSS board to officially approve that we start contributions

at first, we will probably just try to merge every improvement that made it into 3.7.2 to the C++14 branch, plus getting the branch to build without errors for clang-14 on a system with libstdc++-headers from gcc-11 as that is my non-cross development system - if porting anything of this to C++14 would be really tricky, we would then take a closer look at the value/cost proposition of the activity - but of course the closer the C++14 branch will stay to the C++17 mainline, the easier it will be to merge any fixes later on

for later changes from `development`, we will have to look into a bit more detail about what changes really make sense to be merged - for us it will probably be just bugfixes in the main framework, not so much the surrounding services (which we do not use)

**Reply by insi-eb** [reply]
just go give you a heads-up: due to vacation season, the formal approval from my company's OSS board will be delayed for a while

after I get the approval, I plan to create PRs to test-drive the contribution process:
* update spdlog to a newer version that works better with some of the compilers we use
* merge fixes/improvement from current master one after the other until I come to one that doesn't build out-of-the-box in C++14 mode and adapt that one

**Reply by insi-eb** [reply]
one more question: how should the versioning of the C++14 version look like - would 3.6.x work for you or are you following SemVer strictly?

**Reply by insi-eb** [reply]
currently VERSION on the c++14-compliant branch says 4.0.0, which is causing issues

**Reply by jeffdiclemente** [reply]
@insi-eb We do follow SemVer. I think the code was version 3.6.x at the point in which we created the c++14-compliant branch. You are free to change the VERSION on c++14-compliant branch. I'm fine with starting the c++14-compliant branch at 3.6.x and moving forward from there, following SemVer. Thanks.

**Reply by insi-eb** [reply]
as now all improvements and bugfixes from 3.7.2 are in the `c++14-compliant` branch except for the abseil removal and switch to C++17 string_view, and the only commits left out are about adaptation of the changelog and the VERSION file, my next question is:

would you be interested releasing official versions from the C++-14 branch? if you would be interested, what could versioning for this version look like? or would you just keep the branch to be available for whoever is interested in it (in which case, there probably should be something in the main documentation so people can actually find this branch)?

**Reply by jeffdiclemente** [reply]
yes, we can release official versions from the C++14 branch. I have no issue with that. We can release them through the releases page; https://github.com/CppMicroServices/CppMicroServices/releases.

As for versioning the c++14 compliant releases, I'm proposing we start at 3.7.2 and then version the c++14 branch independently from the development branch while still following semantic versioning. This may cause confusion when version #s between the c++14 and development branches overlap however my hope is that the releases will make it clear that the release only supports c++14 and older.

@insi-eb what do you think? feedback is welcome.

**Reply by insi-eb** [reply]
Personally, I'm for making things as explicit as possible. I.e. in this case, releasing a 3.7.2-cpp14 and using independent versions only when the releases are really independent (i.e. next one would be 3.7.4-cpp14 when the branch has caught up to 3.7.4, or 3.7.3-cpp14 if the branch would be somewhere in between .2 and .4).
As far as I can see, SemVer doesn't really support something like that, so the question is how far would you be willing to deviate from "pure" SemVer.

I'm also a bit unsure what the best strategy for versioning the shared objects themselves would be, to make clear that they are potentially (not entirely sure whether or not the C++17 switch makes a difference to the library ABI) not compatible to a C++17-based 3.7.2 - I have not looked at library versioning strategies for a really long time.

Alternatively this could be a 3.6.x version to show that it is more closely related to the last version supporting C++14.

**Reply by jeffdiclemente** [reply]
I looked around for examples of how other projects manage this situation and didn't find much useful information. [abseil ](https://github.com/abseil/abseil-cpp/releases) has the concept of long term support (LTS) where they maintain backwards compatibility with an older C++ standard for a time.

I'm wondering whether we should look at this c++14 branch through the lens of LTS and say that this c++14 branch will be pegged at 3.6.x. My reasoning is that 3.6.x was the last c++14 release, so it makes some sense to me from a code management standpoint. This would mean a deviation from semver when a change is ported to the c++14 branch which isn't a bug fix. However I think I'm ok with that. It's possible that no backwards compatible functionality needs to be pulled into the c++14 branch in the future (although I wouldn't bet on it).

So, I propose we keep the C++14 branch at the 3.6.x version level and any and all code changes will only ever update the patch version. All releases on the github release page will indicate that this release only supports up to C++14. We could always pivot to bump the minor version for the c++14 branch and adopt an explicit versioning scheme (e.g 3.7.2-cpp14) if the need arises in the future.

What do you think?

**Reply by insi-eb** [reply]
I'm fine with staying on a 3.6.x versioning scheme even though features are added to stay merge-compatible with 3.7.2. Internally we also now use the C++14-branch as a 3.6.1
If this is too confusing to users, there still is the explicit versioning scheme as fallback.

I'll create a PR for that as soon as I find some time again - should not take too long.

Btw, is there a reason 3.7.4 has not been merged to master?

**Reply by jeffdiclemente** [reply]
> Btw, is there a reason 3.7.4 has not been merged to master?

no reason other than we haven't gotten to it yet.

**Reply by insi-eb** [reply]
@jeffdiclemente: shall I do anything more for a 3.6.1 release from the C++14 branch or will you handle that (e.g. merging c++14-compliant to the release-3.6 branch and creating a tag)?

**Reply by jeffdiclemente** [reply]
@insi-eb I'll merge c++14-compliant into the release-3.6 branch and create a tag (3.6.1) today.

**Reply by jeffdiclemente** [reply]
@insi-eb All set. I created a new release as well under https://github.com/CppMicroServices/CppMicroServices/releases. Let me know if anything looks wrong and I'll fix it.

**Comment by RobertLauferElektrobit** [comment]
Hi @jeffdiclemente,
could you please create a new release 3.6.4 with content from branch [c++14-compliant](https://github.com/CppMicroServices/CppMicroServices/tree/c%2B%2B14-compliant)?
Last open PR to update CHANGELOG and bump versions is #1190

**Reply by jeffdiclemente** [reply]
@tcormackMW , can you please create a 3.6.4 release of the c++14-compliant branch? Let me know if you need details about how to do this. Thanks.
